# Week 03 Assignment Covid

New types of data and new data science technologies enable new research. These new technologies are technologies such as the ability to combine existing data or the ability to generate synthetic data from existing knowledge. This week casus is based on such research. Data is generated by Synthea's COVID-19 module. The data was constructed using three peer-reviewed publications published in the early stages of the global pandemic, when less was known, along with emerging resources, data, publications, and clinical knowledge. The simulation outputs synthetic Electronic Health Records (EHR), including the daily consumption of Personal Protective Equipment (PPE) and other medical devices and supplies. The Data is stored in separate tables to avoid redundancy, with as a concequence that tables need to be combined and reorganized in dataframes for analysing purpose.

Keywords: merge data, subset data, clean data, generate data

You will learn about combining data with pandas and numpy and you will learn to visualize with bokeh. Concretely, you will preprocess the partly Synthetic Covid data in an appropiate format in order to conduct statistical and visual analysis. Learning objectives

- Combine multiple data sources for analysis
- Read, inspect, clean, reshape data
- Visualize data using bokeh
- Maintain development environment 
- Apply coding standards and FAIR principles
- Reshape the dataset into a format suitable for visual and statistical analysis
- Use widgets to make the plot interactive 
- Use GIS libraries to plot geographical data

Tutorials about combining data: https://github.com/fenna/BFVM22PROG1/blob/main/tutorials/tutorial_combine_data.ipynb

study case combining data:https://github.com/fenna/BFVM22PROG1/blob/main/study_cases/adults_who_binge_drank_in_hot_towns.ipynb


Please add the topics you want to learn about here: https://padlet.com/ffeenstra1/kzh2chaqleq3iovu


Your job is to **visualize the lab values taken for COVID-19 patients of survived versus not survived patients**. 

The assignment consists of 6 parts:

- [part 1: load the data](#0)
     - [Exercise 1.1](#ex-11)
- [part 2: data wrangling](#1)
     - [Exercise 2.1](#ex-21)
- [part 3: more wrangling](#2)
     - [Exercise 3.1](#ex-31)
- [part 4: plot the data](#3)
     - [Exercise 4.1](#ex-41)
- [part 5: plot patient location](#5)
     - [Exercise 5.1](#ex-51)


Part 1 and 4 are mandatory, part 5 is optional (bonus)
Mind you that you cannot copy code without referencing the code. If you copy code you need to be able to explain your code verbally and you will not get the full score. 


## About the data

The data is generated by Synthea's COVID-19 module. The data was constructed using three peer-reviewed publications published in the early stages of the global pandemic, when less was known, along with emerging resources, data, publications, and clinical knowledge. The simulation outputs synthetic Electronic Health Records (EHR), including the daily consumption of Personal Protective Equipment (PPE) and other medical devices and supplies. For this assignment the `conditions`, `patients`, `observations`, `careplans` and `encounters` table will be used. The Data is stored in separate tables to avoid redundancy, with as a concequence that tables need to be combined and reorganized in dataframes for analysing purpose.

Source: Walonoski J, Klaus S, Granger E, Hall D, Gregorowicz A, Neyarapally G, Watson A, Eastman J. Synthea™ Novel coronavirus (COVID-19) model and synthetic data set. Intelligence-Based Medicine. 2020 Nov;1:100007. https://doi.org/10.1016/j.ibmed.2020.100007

Please <a href = "https://synthetichealth.github.io/synthea-sample-data/downloads/10k_synthea_covid19_csv.zip">download</a> the data

#### Covid Patients
Patients are considered Covid patients if they are identified with `CODE` `840539006`


#### Survivors
Patients that had covid and where tested negative after isolation have tested code `94531-1`,  SARS-CoV-2 RNA Pnl Resp NAA+probe (covid-sars test) + a value of `Not detected (qualifier value)`. These patients are considered to be survived covid patients. 

#### Non-Survivors
Patients that did not survived Covid have a `DEATHDATE` which is not null. 


#### Lab values  COVID-19 patients

Patients are monitored for blood and heart conditions once they are admitted in Hospital or under treatment. The lab values of interest are as follow: 

- `48065-7`  Fibrin D-dimer FEU [Mass/volume] in Platelet poor plasma
- `26881-3`   Interleukin 6 [Mass/volume] in Serum or Plasma
- `2276-4` Ferritin [Mass/volume] in Serum or Plasma
- `89579-7` Troponin I.cardiac [Mass/volume] in Serum or Plasma by High sensitivity method
- `731-0` Lymphocytes [#/volume] in Blood by Automated count
- `14804-9` Lactate dehydrogenase [Enzymatic activity/volume] in Serum or Plasma by Lactate to pyruvate reaction


---

<a name='0'></a>
## Part 1: Load the data (20 pt)

Instructions: Load the data of the following files. 
Preferably we read the data not with a hard coded data path but using a config file. See https://fennaf.gitbook.io/bfvm22prog1/data-processing/configuration-files/yaml

- conditions.csv
- patients.csv
- observations.csv
- careplans.csv
- encounters.csv

Get yourself familiar with the data. Create some meaningful overviews. Answer the following questions

1. How many patients are there
2. How many covid-patients are there
3. How many patients do have a 'Hospital admission for isolation' encounter
    
<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
    <ul><li>use a unique dataframe for each file, use a meaningful name</li>
    <li>pandas.read_csv() method can be used to read a csv file</li>
    <li>pandas.DataFrame.head() method is often used to inspect the dataframe</li>
    <li>.unique() returns a list of unique values of a column</li>
</ul>
</details>

<a name='ex-11'></a>
### 1.1 Code your solution

In [417]:
import yaml
with open("config.yaml", 'r') as stream:
    config = yaml.safe_load(stream)
careplan = config['data_careplans']
conditions = config['data_conditions']
patients = config['data_patients']
encounters = config['data_encounters']
observations = config['data_observations']

In [418]:
import pandas as pd
import numpy as np
df_care = pd.read_csv(careplan)
df_cond = pd.read_csv(conditions)
df_pat = pd.read_csv(patients)
df_enco = pd.read_csv(encounters)
df_obsv = pd.read_csv(observations) 

In [419]:
df_care.head()

Id       START        STOP  \
0  fea43343-7312-423f-bb82-b2f5ae71a260  2020-03-01  2020-03-01   
1  cbcade35-42bf-4807-8154-3f7f847221e0  2020-03-01  2020-03-30   
2  51dd78df-2b01-486a-8b33-1fbcd9cec211  2020-02-12  2020-02-26   
3  8aa5055b-cddc-4170-9e31-e71e5552502a  2020-03-13  2020-03-13   
4  976d369a-2b71-488d-ba20-8674fc272be0  2020-03-13  2020-04-14   

                                PATIENT                             ENCOUNTER  \
0  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271  681c380b-3c84-4c55-80a6-db3d9ea12fee   
1  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271  681c380b-3c84-4c55-80a6-db3d9ea12fee   
2  067318a4-db8f-447f-8b6e-f2f61e9baaa5  adedca64-700b-4fb9-82f1-9cbb658abb73   
3  067318a4-db8f-447f-8b6e-f2f61e9baaa5  1ea74a77-3ad3-4948-a9cc-3084462035d6   
4  067318a4-db8f-447f-8b6e-f2f61e9baaa5  1ea74a77-3ad3-4948-a9cc-3084462035d6   

        CODE                                     DESCRIPTION   REASONCODE  \
0  736376001  Infectious disease care plan (record artifact)  840544004.0   
1  736376001  Infectious disease care plan (record artifact)  840539006.0   
2   91251008                      Physical therapy procedure   44465007.0   
3  736376001  Infectious disease care plan (record artifact)  840544004.0   
4  736376001  Infectious disease care plan (record artifact)  840539006.0   

    REASONDESCRIPTION  
0  Suspected COVID-19  
1            COVID-19  
2     Sprain of ankle  
3  Suspected COVID-19  
4            COVID-19

In [420]:
df_cond.head()

START        STOP                               PATIENT  \
0  2019-02-15  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
1  2019-10-30  2020-01-30  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
2  2020-03-01  2020-03-30  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
3  2020-03-01  2020-03-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
4  2020-03-01  2020-03-30  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   

                              ENCOUNTER       CODE         DESCRIPTION  
0  d5ee30a9-362f-429e-a87a-ee38d999b0a5   65363002        Otitis media  
1  8bca6d8a-ab80-4cbf-8abb-46654235f227   65363002        Otitis media  
2  681c380b-3c84-4c55-80a6-db3d9ea12fee  386661006     Fever (finding)  
3  681c380b-3c84-4c55-80a6-db3d9ea12fee  840544004  Suspected COVID-19  
4  681c380b-3c84-4c55-80a6-db3d9ea12fee  840539006            COVID-19

In [421]:
df_pat.head()

Id   BIRTHDATE DEATHDATE          SSN  \
0  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271  2017-08-24       NaN  999-68-6630   
1  067318a4-db8f-447f-8b6e-f2f61e9baaa5  2016-08-01       NaN  999-15-5895   
2  ae9efba3-ddc4-43f9-a781-f72019388548  1992-06-30       NaN  999-27-3385   
3  199c586f-af16-4091-9998-ee4cfc02ee7a  2004-01-09       NaN  999-73-2461   
4  353016ea-a0ff-4154-85bb-1cf8b6cedf20  1996-11-15       NaN  999-60-7372   

     DRIVERS    PASSPORT PREFIX        FIRST        LAST SUFFIX  ...  \
0        NaN         NaN    NaN   Jacinto644     Kris249    NaN  ...   
1        NaN         NaN    NaN      Alva958  Krajcik437    NaN  ...   
2  S99971451  X53218815X    Mr.    Jayson808    Fadel536    NaN  ...   
3  S99956432         NaN    NaN     Jimmie93   Harris789    NaN  ...   
4  S99917327  X58903159X    Mr.  Gregorio366      Auer97    NaN  ...   

                       BIRTHPLACE                         ADDRESS  \
0      Beverly  Massachusetts  US       888 Hickle Ferry Suite 38   
1       Boston  Massachusetts  US             1048 Skiles Trailer   
2  Springfield  Massachusetts  US       1056 Harris Lane Suite 70   
3    Worcester  Massachusetts  US      201 Mitchell Lodge Unit 67   
4              Patras  Achaea  GR  1050 Lindgren Extension Apt 38   

          CITY          STATE           COUNTY     ZIP        LAT        LON  \
0  Springfield  Massachusetts   Hampden County  1106.0  42.151961 -72.598959   
1      Walpole  Massachusetts   Norfolk County  2081.0  42.177370 -71.281353   
2     Chicopee  Massachusetts   Hampden County  1020.0  42.181642 -72.608842   
3     Pembroke  Massachusetts  Plymouth County     NaN  42.075292 -70.757035   
4       Boston  Massachusetts   Suffolk County  2135.0  42.352434 -71.028610   

  HEALTHCARE_EXPENSES HEALTHCARE_COVERAGE  
0             8446.49             1499.08  
1            89893.40             1845.72  
2           577445.86             3528.84  
3           336701.72             2705.64  
4           484076.34             3043.04  

[5 rows x 25 columns]

In [422]:
df_enco.head()

Id                 START  \
0  d5ee30a9-362f-429e-a87a-ee38d999b0a5  2019-02-16T01:02:32Z   
1  6a74fdef-2287-44bf-b9e7-18012376faca  2019-08-02T01:02:32Z   
2  8bca6d8a-ab80-4cbf-8abb-46654235f227  2019-10-31T01:02:32Z   
3  821e57ac-9304-46a9-9f9b-83daf60e9e43  2020-01-31T01:02:32Z   
4  681c380b-3c84-4c55-80a6-db3d9ea12fee  2020-03-02T01:02:32Z   

                   STOP                               PATIENT  \
0  2019-02-16T01:17:32Z  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
1  2019-08-02T01:32:32Z  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
2  2019-10-31T01:17:32Z  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
3  2020-01-31T01:17:32Z  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
4  2020-03-02T01:58:32Z  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   

                           ORGANIZATION                              PROVIDER  \
0  5103c940-0c08-392f-95cd-446e0cea042a  e2c226c2-3e1e-3d0b-b997-ce9544c10528   
1  0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3  87c33fc5-3fd1-3c52-815a-b89a1623bb3a   
2  5103c940-0c08-392f-95cd-446e0cea042a  e2c226c2-3e1e-3d0b-b997-ce9544c10528   
3  0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3  87c33fc5-3fd1-3c52-815a-b89a1623bb3a   
4  fd328395-ab1d-35c6-a2d0-d05a9a79cf11  9c875a09-93e0-39aa-9260-ad264bbdd3fe   

                                  PAYER ENCOUNTERCLASS       CODE  \
0  7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a     outpatient  185345009   
1  7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a       wellness  410620009   
2  7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a     outpatient  185345009   
3  7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a       wellness  410620009   
4  7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a     ambulatory  185345009   

                         DESCRIPTION  BASE_ENCOUNTER_COST  TOTAL_CLAIM_COST  \
0              Encounter for symptom               129.16            129.16   
1       Well child visit (procedure)               129.16            129.16   
2              Encounter for symptom               129.16            129.16   
3       Well child visit (procedure)               129.16            129.16   
4  Encounter for symptom (procedure)               129.16            129.16   

   PAYER_COVERAGE  REASONCODE REASONDESCRIPTION  
0           69.16  65363002.0      Otitis media  
1          129.16         NaN               NaN  
2           69.16  65363002.0      Otitis media  
3          129.16         NaN               NaN  
4           69.16         NaN               NaN

In [423]:
df_obsv.head()

DATE                               PATIENT  \
0  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
1  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
2  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
3  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
4  2019-08-01  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   

                              ENCOUNTER     CODE  \
0  6a74fdef-2287-44bf-b9e7-18012376faca   8302-2   
1  6a74fdef-2287-44bf-b9e7-18012376faca  72514-3   
2  6a74fdef-2287-44bf-b9e7-18012376faca  29463-7   
3  6a74fdef-2287-44bf-b9e7-18012376faca  77606-2   
4  6a74fdef-2287-44bf-b9e7-18012376faca   9843-4   

                                         DESCRIPTION VALUE    UNITS     TYPE  
0                                        Body Height  82.7       cm  numeric  
1  Pain severity - 0-10 verbal numeric rating [Sc...   2.0  {score}  numeric  
2                                        Body Weight  12.6       kg  numeric  
3                  Weight-for-length Per age and sex  86.1        %  numeric  
4               Head Occipital-frontal circumference  46.9       cm  numeric

In [424]:
print("careplan patients:", len(df_care['PATIENT'].unique()))
print("condition patients:", len(df_cond['PATIENT'].unique()))
print("patients id's", len(df_pat['Id'].unique()))
print("encounter patients", len(df_enco['PATIENT'].unique()))
print("observation patients", len(df_obsv['PATIENT'].unique()))

careplan patients: 11771
condition patients: 12165
patients id's 12352
encounter patients 12330
observation patients 12352


1). Total amount of patients is 12352

In [425]:
df_care.groupby('REASONDESCRIPTION')['PATIENT'].count().head(10)

REASONDESCRIPTION
Acute bronchitis (disorder)                    1551
Alzheimer's disease (disorder)                  454
Asthma                                           17
Atopic dermatitis                               174
Attempted suicide - cut/stab                      2
Bullet wound                                      5
COVID-19                                       7014
Child attention deficit disorder                 75
Childhood asthma                                579
Chronic congestive heart failure (disorder)     668
Name: PATIENT, dtype: int64

In [426]:
df_care.groupby('REASONDESCRIPTION')['PATIENT'].count().tail(10)

REASONDESCRIPTION
Second degree burn                         12
Secondary malignant neoplasm of colon      20
Smokes tobacco daily                      144
Sprain of ankle                           155
Sprain of wrist                            48
Suicidal deliberate poisoning               1
Suspected COVID-19                       9106
Tear of meniscus of knee                    7
Third degree burn                           3
Whiplash injury to neck                   321
Name: PATIENT, dtype: int64

In [427]:
df_cond.groupby('DESCRIPTION')['PATIENT'].count().head(30)

DESCRIPTION
Acquired coagulation disorder (disorder)                 127
Acute Cholecystitis                                        3
Acute allergic reaction                                   16
Acute bacterial sinusitis (disorder)                      94
Acute bronchitis (disorder)                              718
Acute deep venous thrombosis (disorder)                  594
Acute pulmonary embolism (disorder)                      587
Acute respiratory distress syndrome (disorder)           234
Acute respiratory failure (disorder)                     862
Acute viral pharyngitis (disorder)                       810
Alcoholism                                                93
Alzheimer's disease (disorder)                           454
Anemia (disorder)                                       3650
Antepartum eclampsia                                      19
Appendicitis                                             581
Asthma                                                    17
At risk for 

2). There are 8820 patients with the condition covid-19
    

There are 7014 patients with a careplan for covid-19 and 9106 with suspected covid-19

In [428]:
df_enco.groupby('DESCRIPTION')['PATIENT'].count()

DESCRIPTION
Admission to intensive care unit (procedure)                                 375
Admission to surgical department                                              59
Admission to thoracic surgery department                                       9
Allergic disorder follow-up assessment                                        34
Allergic disorder initial assessment                                         143
Asthma follow-up                                                             511
Cardiac Arrest                                                               467
Consultation for treatment                                                  2780
Death Certification                                                         1770
Diagnosis of cystic fibrosis using sweat test and gene test                    2
Discussion about treatment (procedure)                                         8
Domiciliary or rest home patient evaluation and management                    61
Drug rehabilitat

3). The amount of patients with hospital admission for isolation is 1867

In [429]:
df_pat.groupby('DEATHDATE')['Id'].unique().count()

1928

In [430]:
df_pat.DEATHDATE.notna().sum()

2352

Extra:

1770 patients have a Death Certification

1928 patients have died due to covid

There are 2352 deathdates registered

### 1.2 Test your solution
The following function needs to be called. You can use this as a test. There are however more meaningful overviews 
you can create. 

In [431]:
def part1(num_pat, num_cov, num_admitted, num_died):
    print(f'There are {num_pat} patients in total')
    print(f'There are {num_cov} covid patients')
    print(f'There are {num_admitted} admitted patients')
    print(f'{num_died} patients died')


### Expected outcome

---

<a name='1'></a>
## Part 2: Data Wrangling: set up the dataframe (30 pt)

In this part we are going to combine data to create a dataframe with values of interest for the lab values analysis. 

We would like a dataframe containing the following information per record (only Covid patients!!!)

- `PATIENT` - the ID of the covid patient
- `days` - the number of days the patient is under observation
- `CODE-Y` - the code of the observation  
- `VALUE` - the lab value of the observation

where only the following observation codes needs to be selected:

- `48065-7`  Fibrin D-dimer FEU [Mass/volume] in Platelet poor plasma
- `26881-3`   Interleukin 6 [Mass/volume] in Serum or Plasma
- `2276-4` Ferritin [Mass/volume] in Serum or Plasma
- `89579-7` Troponin I.cardiac [Mass/volume] in Serum or Plasma by High sensitivity method
- `731-0` Lymphocytes [#/volume] in Blood by Automated count
- `14804-9` Lactate dehydrogenase [Enzymatic activity/volume] in Serum or Plasma by Lactate to pyruvate reaction

The days information is not primarely available and needs to be calculated by substracting observation DATE - START. 

An example of such a dataframe is given below:

In [432]:
#Possible approach:

#Select all the patients with covid from the conditions table
#Combine conditions table (only covid patients) with the patient table into a covid_patient table
#select the only the relevant lab observations from the observations table into a lab_obs table
#merge the covid_patient table with the lab_obs table into a covid_patients_obs table
#clean the covid_patients_obs table (rename columns, select only relevant columns, sort, typecast, add days column)

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
    <ul><li>you can use pandas.DataFrame.merge() to merge dataframes</li>
    <li>df = df[(df.CODE == condition1 | df.CODE == condition1 )] selects rows with CODE of 2 conditional values</li>
    <li>df.DATE - df.START return days if DATE and START are datetime format</li>
    <li>pd.to_datetime() can be used to typecast to datetime</li>
</ul>
</details>

<a name='ex-21'></a>
### 2.1 Code your solution

In [433]:
# Sort on only covid patients
df_covid = df_cond[df_cond['DESCRIPTION'] == 'COVID-19']


In [434]:
# Sort only on the wanted codes
df_code = df_obsv[(df_obsv['CODE'] == '48065-7')| (df_obsv['CODE'] == '26881-3')| (df_obsv.CODE == '2276-4')| (df_obsv.CODE == '89579-7')| (df_obsv.CODE == '731-0')| (df_obsv.CODE == '14804-9')]

In [435]:
df_code

DATE                               PATIENT  \
273      2020-02-19  f58bf921-cba1-475a-b4f8-dc6fa3b8f89c   
292      2020-02-19  f58bf921-cba1-475a-b4f8-dc6fa3b8f89c   
293      2020-02-19  f58bf921-cba1-475a-b4f8-dc6fa3b8f89c   
294      2020-02-19  f58bf921-cba1-475a-b4f8-dc6fa3b8f89c   
295      2020-02-19  f58bf921-cba1-475a-b4f8-dc6fa3b8f89c   
...             ...                                   ...   
1658604  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   
1658623  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   
1658624  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   
1658625  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   
1658626  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   

                                    ENCOUNTER     CODE  \
273      e97e8d37-7497-4c13-98fd-a4a45655c0bb    731-0   
292      e97e8d37-7497-4c13-98fd-a4a45655c0bb  48065-7   
293      e97e8d37-7497-4c13-98fd-a4a45655c0bb   2276-4   
294      e97e8d37-7497-4c13-98fd-a4a45655c0bb  89579-7   
295      e97e8d37-7497-4c13-98fd-a4a45655c0bb  14804-9   
...                                       ...      ...   
1658604  57d8ff2e-b92c-4fb5-bbf4-d7d5f23382b4    731-0   
1658623  57d8ff2e-b92c-4fb5-bbf4-d7d5f23382b4  48065-7   
1658624  57d8ff2e-b92c-4fb5-bbf4-d7d5f23382b4   2276-4   
1658625  57d8ff2e-b92c-4fb5-bbf4-d7d5f23382b4  89579-7   
1658626  57d8ff2e-b92c-4fb5-bbf4-d7d5f23382b4  14804-9   

                                               DESCRIPTION  VALUE    UNITS  \
273      Lymphocytes [#/volume] in Blood by Automated c...    1.1  10*3/uL   
292      Fibrin D-dimer FEU [Mass/volume] in Platelet p...    0.4    ug/mL   
293              Ferritin [Mass/volume] in Serum or Plasma  332.4     ug/L   
294      Troponin I.cardiac [Mass/volume] in Serum or P...    2.3    pg/mL   
295      Lactate dehydrogenase [Enzymatic activity/volu...  223.9      U/L   
...                                                    ...    ...      ...   
1658604  Lymphocytes [#/volume] in Blood by Automated c...    0.9  10*3/uL   
1658623  Fibrin D-dimer FEU [Mass/volume] in Platelet p...    0.5    ug/mL   
1658624          Ferritin [Mass/volume] in Serum or Plasma  525.2     ug/L   
1658625  Troponin I.cardiac [Mass/volume] in Serum or P...    3.0    pg/mL   
1658626  Lactate dehydrogenase [Enzymatic activity/volu...  247.9      U/L   

            TYPE  
273      numeric  
292      numeric  
293      numeric  
294      numeric  
295      numeric  
...          ...  
1658604  numeric  
1658623  numeric  
1658624  numeric  
1658625  numeric  
1658626  numeric  

[73918 rows x 8 columns]

---

In [436]:
df_covid

START        STOP                               PATIENT  \
4       2020-03-01  2020-03-30  f0f3bc8d-ef38-49ce-a2bd-dfdda982b271   
11      2020-03-13  2020-04-14  067318a4-db8f-447f-8b6e-f2f61e9baaa5   
23      2020-03-11  2020-04-15  ae9efba3-ddc4-43f9-a781-f72019388548   
30      2020-03-02  2020-04-07  199c586f-af16-4091-9998-ee4cfc02ee7a   
37      2020-03-02  2020-03-18  353016ea-a0ff-4154-85bb-1cf8b6cedf20   
...            ...         ...                                   ...   
114458  2020-03-07  2020-03-17  534ed6ac-1178-47ce-a0c4-ecf779613143   
114474  2020-03-08  2020-03-16  c9699449-7a8b-400a-8e86-fab6aa7134cb   
114490  2020-02-29  2020-03-18  4e5d67bf-0f2e-4158-af57-413e276c84dc   
114528  2020-03-02  2020-04-02  435be43f-cb57-442f-aeb2-51018b030ed0   
114543  2020-03-04  2020-04-01  2712205f-755e-4897-acb3-926895b7d635   

                                   ENCOUNTER       CODE DESCRIPTION  
4       681c380b-3c84-4c55-80a6-db3d9ea12fee  840539006    COVID-19  
11      1ea74a77-3ad3-4948-a9cc-3084462035d6  840539006    COVID-19  
23      eeab7c2d-71ba-4e04-af16-87a01dce7d54  840539006    COVID-19  
30      8333efdf-f7bf-43bb-b73f-2b663d14c1ad  840539006    COVID-19  
37      8d502eec-afc0-4639-b298-a152a16d492a  840539006    COVID-19  
...                                      ...        ...         ...  
114458  852042ad-2151-4039-9fac-a28858f66968  840539006    COVID-19  
114474  9c7a5b12-a07d-406a-95b3-d7454fc59468  840539006    COVID-19  
114490  b179344f-f271-4f52-acca-427ae71b23ac  840539006    COVID-19  
114528  a3232b65-d975-45d1-a9bd-1e89704d7f58  840539006    COVID-19  
114543  a18d0f16-bc12-4897-8844-8c3ebc2464fc  840539006    COVID-19  

[8820 rows x 6 columns]

In [437]:

left = pd.DataFrame({'PATIENT': df_covid['PATIENT'], 'START': df_covid['START']})
left 


PATIENT       START
4       f0f3bc8d-ef38-49ce-a2bd-dfdda982b271  2020-03-01
11      067318a4-db8f-447f-8b6e-f2f61e9baaa5  2020-03-13
23      ae9efba3-ddc4-43f9-a781-f72019388548  2020-03-11
30      199c586f-af16-4091-9998-ee4cfc02ee7a  2020-03-02
37      353016ea-a0ff-4154-85bb-1cf8b6cedf20  2020-03-02
...                                      ...         ...
114458  534ed6ac-1178-47ce-a0c4-ecf779613143  2020-03-07
114474  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-08
114490  4e5d67bf-0f2e-4158-af57-413e276c84dc  2020-02-29
114528  435be43f-cb57-442f-aeb2-51018b030ed0  2020-03-02
114543  2712205f-755e-4897-acb3-926895b7d635  2020-03-04

[8820 rows x 2 columns]

In [438]:
right = pd.DataFrame({'PATIENT': df_code['PATIENT'], 'DATE':df_code['DATE'], 'CODE': df_code['CODE'], 'VALUE':df_code['VALUE'], 'UNITS':df_code['UNITS']})
right

PATIENT        DATE     CODE  VALUE  \
273      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c  2020-02-19    731-0    1.1   
292      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c  2020-02-19  48065-7    0.4   
293      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c  2020-02-19   2276-4  332.4   
294      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c  2020-02-19  89579-7    2.3   
295      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c  2020-02-19  14804-9  223.9   
...                                       ...         ...      ...    ...   
1658604  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-16    731-0    0.9   
1658623  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-16  48065-7    0.5   
1658624  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-16   2276-4  525.2   
1658625  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-16  89579-7    3.0   
1658626  c9699449-7a8b-400a-8e86-fab6aa7134cb  2020-03-16  14804-9  247.9   

           UNITS  
273      10*3/uL  
292        ug/mL  
293         ug/L  
294        pg/mL  
295          U/L  
...          ...  
1658604  10*3/uL  
1658623    ug/mL  
1658624     ug/L  
1658625    pg/mL  
1658626      U/L  

[73918 rows x 5 columns]

In [439]:
df_new = pd.merge(left, right, how='inner', left_on=['PATIENT'], right_on=['PATIENT'])

In [440]:
df_new['START'] = pd.to_datetime(df_new['START'],format='%Y-%m-%d')
df_new['DATE'] = pd.to_datetime(df_new['DATE'],format='%Y-%m-%d') 

In [441]:
df_new['DAYS'] = df_new['DATE'] - df_new['START']

In [442]:
df_new.drop('DATE', axis=1, inplace = True)
df_new.drop('START', axis=1, inplace = True)

In [443]:
df = df_new[['PATIENT', 'DAYS', 'CODE','VALUE', 'UNITS']]

In [444]:
df

PATIENT   DAYS     CODE  VALUE    UNITS
0      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days    731-0    1.1  10*3/uL
1      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days  48065-7    0.4    ug/mL
2      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days   2276-4  332.4     ug/L
3      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days  89579-7    2.3    pg/mL
4      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days  14804-9  223.9      U/L
...                                     ...    ...      ...    ...      ...
73913  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days    731-0    0.9  10*3/uL
73914  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days  48065-7    0.5    ug/mL
73915  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days   2276-4  525.2     ug/L
73916  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days  89579-7    3.0    pg/mL
73917  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days  14804-9  247.9      U/L

[73918 rows x 5 columns]

In [445]:
df_test = df[df['PATIENT'] == '00079a57-24a8-430f-b4f8-a1cf34f90060']

In [446]:
df_test.head(10)

PATIENT   DAYS     CODE  VALUE    UNITS
48607  00079a57-24a8-430f-b4f8-a1cf34f90060 0 days    731-0    1.1  10*3/uL
48608  00079a57-24a8-430f-b4f8-a1cf34f90060 0 days  48065-7    0.4    ug/mL
48609  00079a57-24a8-430f-b4f8-a1cf34f90060 0 days   2276-4  489.5     ug/L
48610  00079a57-24a8-430f-b4f8-a1cf34f90060 0 days  89579-7    3.4    pg/mL
48611  00079a57-24a8-430f-b4f8-a1cf34f90060 0 days  14804-9  241.8      U/L
48612  00079a57-24a8-430f-b4f8-a1cf34f90060 1 days    731-0    1.1  10*3/uL
48613  00079a57-24a8-430f-b4f8-a1cf34f90060 2 days    731-0    1.0  10*3/uL
48614  00079a57-24a8-430f-b4f8-a1cf34f90060 2 days  48065-7    0.3    ug/mL
48615  00079a57-24a8-430f-b4f8-a1cf34f90060 2 days   2276-4  497.3     ug/L
48616  00079a57-24a8-430f-b4f8-a1cf34f90060 2 days  89579-7    2.0    pg/mL

<a name='2'></a>
## Part 3: Data Wrangling, split into survived and not survived (10 pt)

Now we have the required data we would like to split the data into survived and not survived. First we fetch all the ids of the survived and deceased patients. We can use these ids to select the records of the survived patients and the patients that did not survived

Your job is to split the data into survived and not survived records. There are multiple ways to do this. One way is the  `.isin()` method

In [447]:
#the following code is given, RUN THIS CELL
#get survived and deceased ids
# completed_isolation_patients = care_plans[(care_plans.CODE == 736376001) & (care_plans.STOP.notna()) \
#                                           & (care_plans.REASONCODE == 840539006)].PATIENT
# negative_covid_patient_ids = observations[(observations.CODE == '94531-1') \
#                                           & (observations.VALUE == 'Not detected (qualifier value)')].PATIENT.unique()
# survivor_ids = np.union1d(completed_isolation_patients, negative_covid_patient_ids)
# deceased_ids = patients[patients.DEATHDATE.notna()].Id

In [448]:
#the following code is given, RUN THIS CELL
#get survived and deceased ids
completed_isolation_patients = df_care[(df_care.CODE == 736376001) & (df_care.STOP.notna()) \
                                          & (df_care.REASONCODE == 840539006)].PATIENT
negative_covid_patient_ids = df_obsv[(df_obsv.CODE == '94531-1') \
                                          & (df_obsv.VALUE == 'Not detected (qualifier value)')].PATIENT.unique()
survivor_ids = np.union1d(completed_isolation_patients, negative_covid_patient_ids)
deceased_ids = df_pat[df_pat.DEATHDATE.notna()].Id

<a name='ex-31'></a>
### 3.1 Code your solution

In [449]:
df['deceased'] = df['PATIENT'].isin(deceased_ids)
df['survived'] = df['PATIENT'].isin(survivor_ids)
print("patient records that have survived:", len(df[survived]))
print("patient records that are deceased:", len(df[deceased]))

patient records that have survived: 57303
patient records that are deceased: 16793


/tmp/ipykernel_3838578/270349473.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['deceased'] = df['PATIENT'].isin(deceased_ids)
/tmp/ipykernel_3838578/270349473.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['survived'] = df['PATIENT'].isin(survivor_ids)


### 3.2 Test your solution

In [450]:
def test3(survived, died):
    print(f'patients records survived: {survived}, patients records deceased {died}')
#call the test3

#### Expected outcome

---

<a name='3'></a>
## Part 4: Plot the data (20 pt)

Create plots with the lab data, for each code one plot. Separate the survivors and the deceased by color. An example of such a plot is given below. You can create 6 plots in one grid (for each code one plot) or use a widget (for instance a drop down menu widget) to select a lab CODE. Plot on the x-axis the days, on the y-axis the VALUE. Use proper labels, titles and legends.

<img src="../images/week3_plot.png" width="500" height="500"/>

<a name='ex-41'></a>
### 4.1 Code your solution

In [451]:
from bokeh.io import output_notebook
from bokeh.plotting import figure, show
from bokeh.layouts import gridplot
from bokeh.io import output_notebook
from bokeh.plotting import ColumnDataSource
from bokeh.models import Legend
from bokeh.models import DatetimeTickFormatter

output_notebook()

Loading BokehJS ...

In [460]:
df.dtypes


PATIENT              object
DAYS        timedelta64[ns]
CODE                 object
VALUE                object
UNITS                object
deceased               bool
survived               bool
dtype: object

In [452]:
#df['VALUE'] = df['VALUE'].astype(int)

In [453]:
df_feu = df[df['CODE'] == '48065-7']

In [459]:
df_feu

PATIENT   DAYS     CODE VALUE  UNITS  \
1      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 0 days  48065-7   0.4  ug/mL   
7      f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 2 days  48065-7   0.3  ug/mL   
13     f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 4 days  48065-7   0.2  ug/mL   
19     f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 6 days  48065-7   0.5  ug/mL   
25     f58bf921-cba1-475a-b4f8-dc6fa3b8f89c 8 days  48065-7   0.5  ug/mL   
...                                     ...    ...      ...   ...    ...   
73890  c9699449-7a8b-400a-8e86-fab6aa7134cb 0 days  48065-7   0.2  ug/mL   
73896  c9699449-7a8b-400a-8e86-fab6aa7134cb 2 days  48065-7   0.2  ug/mL   
73902  c9699449-7a8b-400a-8e86-fab6aa7134cb 4 days  48065-7   0.4  ug/mL   
73908  c9699449-7a8b-400a-8e86-fab6aa7134cb 6 days  48065-7   0.5  ug/mL   
73914  c9699449-7a8b-400a-8e86-fab6aa7134cb 8 days  48065-7   0.5  ug/mL   

       deceased  survived  
1         False      True  
7         False      True  
13        False      True  
19        False      True  
25        False      True  
...         ...       ...  
73890     False      True  
73896     False      True  
73902     False      True  
73908     False      True  
73914     False      True  

[12809 rows x 7 columns]

In [455]:
p = figure(title='Fibrin D-dimer FEU', x_axis_type="datetime")
p.xaxis.axis_label = 'Days'
p.yaxis.axis_label = 'ug/mL'

#source = ColumnDataSource(df_feu)

#p.circle(x='DAYS', y='VALUE', source=source, color='blue', size=10, legend_label='deceased')
#p.circle(df_feu.DAYS, df_feu.VALUE, color=df_feu.deceased, size=10, fill_color=None, legend_label='survived')

p.legend.click_policy='hide'

#legend = Legend(items=[('deceased',[d]), ('survived',[s])], location='center')
#p.add_layout(legend, 'right')
show(p)

/commons/conda/dsls/lib/python3.10/site-packages/bokeh/models/plots.py:815: UserWarning: 
You are attempting to set `plot.legend.click_policy` on a plot that has zero legends added, this will have no effect.

Before legend properties can be set, you must add a Legend explicitly, or call a glyph method with a legend parameter set.

  warnings.warn(_LEGEND_EMPTY_WARNING % attr)


In [456]:
# 48065-7 Fibrin D-dimer FEU [Mass/volume] in Platelet poor plasma
# 26881-3 Interleukin 6 [Mass/volume] in Serum or Plasma
# 2276-4 Ferritin [Mass/volume] in Serum or Plasma
# 89579-7 Troponin I.cardiac [Mass/volume] in Serum or Plasma by High sensitivity method
# 731-0 Lymphocytes [#/volume] in Blood by Automated count
# 14804-9 Lactate dehydrogenase [Enzymatic activity/volume] in Serum or Plasma by Lactate to pyruvate reaction

In [457]:
import matplotlib
import matplotlib.pyplot as plt
plt.scatter(df['DAYS'], df['VALUE'],)

<a name='4'></a>
## Part 5: Plot the location of the patients (10 pt)

This is a bonus part. Can you plot the patients location on a map? See also 
https://docs.bokeh.org/en/latest/docs/user_guide/geo.html 

You can use either package folium or geopandas. You need the Latitude and Longitude information from the patient tabel


<a name='ex-51'></a>
### 5.1 Code your solution

In [ ]:
#YOUR CODE HERE